Stepped reservation process

ryanmitchell opened the pull request:
WIP at this point, but to give you a quick idea of what I was thinking.

Although its now multi-step, the benefits of this are we now show the right dates for each location, and the right times for the date selected.

If you're happy with the approach, then next I would be working on styling each of the steps (e.g. location list, calendar/date selection).

I had also wondered if it would be better to move reservations to an URL like {location_slug}/reservations like menus is?

ryanmitchell commented:
An alternative would be to switch to a calendar view similar to how Resdiary handle it. Changing the guest number, time etc would update the calendar dates through a JS request.

![Screenshot 2020-09-07 at 12 23 14](https://user-images.githubusercontent.com/51899/92382685-ea392c00-f104-11ea-89b4-24fed0eb3985.png)

ryanmitchell commented:
having thought it over some more I think the calendar view is the way to do, its much more user friendly.

So here's an update with that approach in place. 

Some thoughts at this point:
1. Do we need the time picker step - I left it in but we could easily incorporate the times into the drop down on the calendar step.
2. I think it would be great to bring this in under the restaurant urls, as it cuts out the first step.

Let me know what you think, I'll hold off doing any more until you come back to me.

sampoyigi commented:
Sorry somehow lost this PR in my notifications.

> If you're happy with the approach, then next I would be working on styling each of the steps (e.g. location list, calendar/date selection).

Yep, go-ahead!

> I had also wondered if it would be better to move reservations to an URL like {location_slug}/reservations like menus is?

Agreed

> Do we need the time picker step - I left it in but we could easily incorporate the times into the drop down on the calendar step.

Love the idea as long as the user experience on mobile is smooth.

ryanmitchell commented:
Great, I'll work on it this week.
Whats the best approach to enable the {location_slug}/reservations url?

ryanmitchell commented:
Ok, I've pushed an update here and its working, though testing and refinements will no doubt be needed. Its working well on mobile and feels more intuitive (to me at least).

The permalink on the theme reservation layout needs changed to `permalink: /:location?local/reservation` for this to function.

If you notice any issues or want any changes let me know.

ryanmitchell commented:
@sampoyigi just in case you've missed this - let me know your thoughts.

sampoyigi commented:
> feels more intuitive

1000%, love the layout.

> Whats the best approach to enable the {location_slug}/reservations URL?

How about add a location select box above the reserve a table form and then reload the page based on the selected location. For example, when Location 1 is selected the page redirects to `/reservation/location-1` and renders the picker_form

ryanmitchell commented:
So I made it work at {slug}/reservations by amending the url in the reservation partial, I referenced it in one of the commments. 

This feels natural to me, that way you would choose a location from the main locations list and there would be a link in the Nav to reserve just like menus?

sampoyigi commented:
Yea I get that with using the main local search box, but that's more for the order and we also want to use to be able to choose a different restaurant from the same page if they change their mind instead of navigating back to the homepage.

ryanmitchell commented:
Ok, location selection is back.

Do we need a 'loading' sort of overlay when you change option to indicate something has happened? If so is there a style for these anywhere?

sampoyigi commented:
Sorry, lost track of this one. We could instead disable inputs when loading or use something similar on the checkout page when choosing payment.

ryanmitchell commented:
You and me both :) 
I've added a progress indicator over the top when loading - let me know if you think thats not enough.

sampoyigi commented on components/Booking.php:
```diff
@@ -279,7 +339,7 @@ public function processPickerForm()
         if (!$this->validatePasses($data, $this->createRules('picker')))
             return;
 
-        $this->pickerStep = array_get($data, 'sdateTime') ? 'info' : 'timeslot';
+        $this->pickerStep = 'info';
```
Can we bring back the timeslot partial? Not sure why you removed it but OpenTable and resdiary have something similar.

ryanmitchell replied:
The reason for leaving it out was that all the timeslots are now on the calendar page - so you'd be asking someone to choose it to then choose it again. I haven't looked at the changes you've made so that may no longer be the case...

sampoyigi replied:
Yea I get you but the idea is to give the customer more options in case they want a different time or maybe the time they selected is fully booked ex. customer selects 6:30 which is fully booked but 7:00 is available.

ryanmitchell replied:
Np. Will work on bringing it back next week.

sampoyigi replied:
Thanks!

sampoyigi commented on components/booking/picker_form.blade.php:
```diff
@@ -6,56 +6,77 @@
+                <label for="time">@lang('igniter.reservation::default.label_time')</label>
+                @partial('@input_time')
+            </div>
+            @if (count($timeOptions))
+                <div class="form-group">
+                    <button
+                        type="submit"
+                        class="btn btn-primary btn-block"
+                    >@lang('igniter.reservation::default.button_find_table')</button>
+                </div>
+            @endif
```
Can we use something like `$this->manager->getSchedule()->isOpenAt($dateTime)` here instead?

ryanmitchell replied:
Again havent looked at the changes thoroughly, so this may no longer apply, but the idea was to only show the button if there were timeslots still available on that day. It may be that the restaurant is open but has no slots available?

sampoyigi replied:
Ok, that makes sense.

ryanmitchell commented on components/Booking.php:
```diff
@@ -249,14 +275,25 @@ public function getReservation()
+        $this->page['nextOpen'] = Carbon::parse($this->manager->getSchedule()->getOpenTime());
+        $this->page['timeOptions'] = $this->getTimeSlots();
+        $this->page['disabledDaysOfWeek'] = $this->getDisabledDaysOfWeek();
+        $this->page['disabledDates'] = $this->getDisabledDates();
```
For disabled dates? It may be worth leaving it out until its supported by the admin panel, I only added it as it was part of the v3 milestone list.

sampoyigi replied:
Alright, we can leave it then.

ryanmitchell commented:
Ok this should be timeslots back and working. As I had amended the timeslots function i added a new one (reducedTimeslots). I've also changed up the HTML a bit to be a button rather than a link. Do you not think all the forms should POST rather than GET? I didn't want to change that in case you had a reason for it.

My only other thought is the location list should be filtered by those that accept reservations - I had that in but you removed it - do you want all locations to be displayed even if they dont accept reservations?

sampoyigi commented:
The GET method is for SEO and to make the booking form sharable. Got the idea from OpenTable.

Erm can't remember removing the code, must have thought it was for something else. Feel free to add it back.

ryanmitchell commented:
Ok the location filtering is back in now.
Let me know what you think.

sampoyigi commented:
Thank you, I'll test and let you know.

BreakSecurity commented:
How can I test all those feature? How can I setup an environment to start doing pull requests? Thanks guys

ryanmitchell commented:
You can download the branch from github as a zip.  

You can also set up git on your computer and then switch to branches which pulls down the code for that branch. 

https://github.com/yui/yui3/wiki/Set-Up-Your-Git-Environment
Might help

> On 8 Nov 2020, at 12:58, BreakSecurity <notifications@github.com> wrote:
> 
> ﻿
> How can I test all those feature? How can I setup an environment to start doing pull requests? Thanks guys
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

sampoyigi commented:
I think we're done here?

ryanmitchell commented:
Yep I think it’s good to merge

sampoyigi commented:
@ryanmitchell Much appreciation for your contributions ❤️
